File: 2025/solutions/day_4.go
package solutions

import (
	"fmt"
	"strconv"
	"strings"
)

var dirs = [8][2]int{
	{-1, -1},
	{-1, 0},
	{-1, 1},
	{0, -1},
	{0, 1},
	{1, -1},
	{1, 0},
	{1, 1},
}

type Day4 struct{}

func (d Day4) Part1(input string) string {
	grid := getGrid(input)

	var accessibleRolls int

	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == '@' && neighbors(grid, i, j) < 4 {
				fmt.Printf("rolls at row %d, column %d is Accessible\n", i+1, j+1)
				accessibleRolls++
			}
		}
	}

	return strconv.Itoa(accessibleRolls)
}

func (d Day4) Part2(input string) string {
	grid := getGrid(input)
	var sum int

	for {
		var accessibleRolls int
		for i := range grid {
			for j := range grid[i] {
				if grid[i][j] == '@' && neighbors(grid, i, j) < 4 {
					fmt.Printf("rolls at row %d, column %d is Accessible\n", i+1, j+1)
					grid[i][j] = '.'
					accessibleRolls++
				}
			}
		}
		if accessibleRolls == 0 {
			break
		}
		sum += accessibleRolls
	}

	return strconv.Itoa(sum)
}

func neighbors(grid [][]byte, row, col int) int {
	var count int

	for _, dir := range dirs {
		r := row + dir[0]
		c := col + dir[1]

		if r >= 0 && r < len(grid) &&
			c >= 0 && c < len(grid[r]) &&
			grid[r][c] == '@' {
			count++
		}
	}

	return count
}

func getGrid(input string) [][]byte {
	lines := strings.Split(input, "\n")
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		grid[i] = []byte(line)
	}
	return grid
}

File: src/solutions/year_2025/day_4.rs
use crate::Answer;

#[derive(Default)]
pub struct Day4;

impl Answer for Day4 {
    fn part_1(&self, input: &str) -> String {
        let mut grid = get_grid(input);
        let h = grid.len();
        let w = grid[0].len();

        get_rolls(&mut grid, h, w, |_| {}).to_string()
    }

    fn part_2(&self, input: &str) -> String {
        let mut grid = get_grid(input);
        let h = grid.len();
        let w = grid[0].len();
        let mut sum = 0;

        loop {
            let accessible_rolls = get_rolls(&mut grid, h, w, |roll| *roll = b'.');
            if accessible_rolls == 0 {
                break;
            }
            sum += accessible_rolls
        }
        sum.to_string()
    }
}

const DIRS: [(isize, isize); 8] = [
    (-1, -1),
    (-1, 0),
    (-1, 1),
    (0, -1),
    (0, 1),
    (1, -1),
    (1, 0),
    (1, 1),
];

fn get_rolls<F>(grid: &mut [Vec<u8>], h: usize, w: usize, mut remove_rolls: F) -> u32
where
    F: FnMut(&mut u8),
{
    let mut count = 0;
    for i in 0..h {
        for j in 0..w {
            if grid[i][j] == b'@' && neighbors(grid, i, j, h, w) < 4 {
                count += 1;
                remove_rolls(&mut grid[i][j]);
            }
        }
    }
    count
}

fn neighbors(grid: &[Vec<u8>], i: usize, j: usize, h: usize, w: usize) -> u8 {
    let mut count = 0;
    for (di, dj) in DIRS {
        let ni = i as isize + di;
        let nj = j as isize + dj;
        if ni >= 0
            && nj >= 0
            && ni < h as isize
            && nj < w as isize
            && grid[ni as usize][nj as usize] == b'@'
        {
            count += 1;
        }
    }
    count
}

fn get_grid(input: &str) -> Vec<Vec<u8>> {
    input.lines().map(|line| line.as_bytes().to_vec()).collect()
}

#[cfg(test)]
mod tests {
    use crate::{Answer, solutions::year_2025::day_4::Day4};

    const TEST: &str = r"..@@.@@@@.
@@@.@.@.@@
@@@@@.@.@@
@.@@@@..@.
@@.@@@@.@@
.@@@@@@@.@
.@.@.@.@@@
@.@@@.@@@@
.@@@@@@@@.
@.@.@@@.@.";

    #[test]
    fn test_part_1() {
        let day = Day4;
        assert_eq!(day.part_1(TEST), "13");
    }
    #[test]
    fn test_part_2() {
        let day = Day4;
        assert_eq!(day.part_2(TEST), "43");
    }
}
